librad/src/git/ext/reference/name.rs

```diff
@@ -0,0 +1,504 @@
+///
+/// Note that refspec patterns (eg. "refs/heads/*") are not allowed (see
+/// [`RefspecPattern`]), and that the maximum length of the name is 1024 bytes.
+///
+/// [`git-check-ref-format`]: https://git-scm.com/docs/git-check-ref-format
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd, Hash)]
+pub struct RefLike(PathBuf);
+
+impl RefLike {
+    pub fn as_str(&self) -> &str {
+        self.into()
+    }
+
+    pub fn percent_encode(&self) -> PercentEncode {
+        /// https://url.spec.whatwg.org/#fragment-percent-encode-set
+        const FRAGMENT_PERCENT_ENCODE_SET: &percent_encoding::AsciiSet =
+            &percent_encoding::CONTROLS
+                .add(b' ')
+                .add(b'"')
+                .add(b'<')
+                .add(b'>')
+                .add(b'`');
+
+        /// https://url.spec.whatwg.org/#path-percent-encode-set
+        const PATH_PERCENT_ENCODE_SET: &percent_encoding::AsciiSet = &FRAGMENT_PERCENT_ENCODE_SET
+            .add(b'#')
+            .add(b'?')
+            .add(b'{')
+            .add(b'}');
+
+        percent_encoding::utf8_percent_encode(self.as_str(), PATH_PERCENT_ENCODE_SET)
+    }
+}
+
+impl Deref for RefLike {
+    type Target = Path;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl TryFrom<&str> for RefLike {
+    type Error = Error;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        normalize_name(
+            s,
+            git2::ReferenceFormat::ALLOW_ONELEVEL | git2::ReferenceFormat::REFSPEC_SHORTHAND,
+        )
+        .map(Self)
+    }
+}
+
+impl TryFrom<&[u8]> for RefLike {
+    type Error = Error;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        str::from_utf8(bytes)
+            .or(Err(Error::Utf8))
+            .and_then(Self::try_from)
+    }
+}
+
+impl FromStr for RefLike {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::try_from(s)
+    }
+}
+
+impl TryFrom<String> for RefLike {
+    type Error = Error;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        Self::try_from(s.as_str())
+    }
+}
+
+impl TryFrom<PathBuf> for RefLike {
+    type Error = Error;
+
+    fn try_from(path: PathBuf) -> Result<Self, Self::Error> {
+        path.to_str().ok_or(Error::Utf8).and_then(Self::try_from)
+    }
+}
+
+impl TryFrom<&Path> for RefLike {
+    type Error = Error;
+
+    fn try_from(path: &Path) -> Result<Self, Self::Error> {
+        path.to_str().ok_or(Error::Utf8).and_then(Self::try_from)
+    }
+}
+
+impl<'a> From<&'a RefLike> for &'a str {
+    fn from(reflike: &'a RefLike) -> &'a str {
+        reflike
+            .0
+            .to_str()
+            .expect("cannot be constructed from invalid utf8")
+    }
+}
+
+/// A [`RefLike`] without a "refs/" prefix.
+///
+/// Conversion functions strip the first **two** path components iff the path
+/// starts with `refs/`.
```

Comment:
Just to clarify, if it's taking the first two components does that mean it's stripping `refs/<next_level>`? For example, `refs/heads/mistress` becomes `mistress` and `refs/remotes/origin/of-life` becomes `origin/of-life`.

Reply:
Ya, I mean it makes sense in that I realise git does this a lot. I think since we're working with the raw strings it's fine to do this. Maybe just calling it out in the docs with examples will make it more clear.

Reply:
Made you some :guide_dog: s

Reply:
![image](https://user-images.githubusercontent.com/3503528/96148558-465f4f00-0f00-11eb-8884-c5919a236aa9.png)